### PR TITLE
Fix max preview, some resizing and caching issues and force preview providers to resize their previews properly

### PR DESCRIPTION
**Pull request by oparoz**
## TL;DR

This PR:
- [introduces a method in OC_Image](https://github.com/owncloud/core/pull/16382/files#diff-8590d19664dac0d5de93639c3e21373cR984) which doesn't stretch images when trying to make them fit in a box
- adds the method to all key providers so that they can do their job, as expected by the Preview class
- [improves the caching mechanism of Preview](https://github.com/oparoz/core/commit/d4483fd25c4c55ad8e188fab1b8b2c5866e4ee94) in order to reduce I/O and to avoid filling the available disk space
- fixes some long standing issues
- **contains mostly tests**
## Description

https://github.com/owncloud/core/issues/13607 describes what needed to be done **at the provider level** in order to support the max preview feature, but unfortunately, only the bitmap provider was fixed until now and this led to some max previews being cropped (https://github.com/owncloud/core/pull/16355).

Since all the other providers end up with an \OC_Image object, they can simply use one of the available methods to make sure the preview they create fits in the asked dimension. I decided to call that method from within each provider since they're responsible to provide a preview which matches the requirements.

Unfortunately, since none of the available method does the job, a new one had to be introduced to make sure a smaller preview wouldn't be stretched to fit in the max dimensions.

One test per group of provider is provided.

Also, since when working on this I encountered many bugs, both introduced by the max preview and through the years, I've decided to **write tests (41 x 4 samples)** for the Preview class to try and detect most of the problems and to fix them.

This branch will now also fix
- [x] Never cache previews larger than the max preview https://github.com/owncloud/core/issues/16545
- [x] Don't store the exact same preview every time we check for its existence https://github.com/owncloud/core/issues/16544
- [x]  Preview can't resize wide images properly #14712 

**Enjoy and hopefully this will be working well for all of you!**

I'm also open to discussing testing strategies as this is not easy to do properly and requires large data samples.
## Test data

I suggest you upload at least these 4 pictures from the tests/data folder to your test folder:
- `testimage.eps` | standard eps file, converted by ImageMagick
- `testimage.jpg` | normal jpg with some text in it
- `testimage-wide.png` | panoramic picture
- `testimageblah.gif` | small animated gif

They were chosen because they trigger different scenarios in the resizing mechanism. You're welcome to test with your libraries as well, just make sure you've deleted the thumbnail folder before proceeding.

Your configuration should have the max preview dimensions set at 756x756 if you want to be able to really test the various restrictions the resizing can encouter.
If you want to test upscaling, then you should have a max_factor of above 1.0 in your configuration
#### URLs

**No aspect ratio**
`/index.php/core/preview.png?file=testimage.jpg&x=756&y=756&forceIcon=1`
**With ratio**
`/index.php/core/preview.png?file=testimage.jpg&x=756&y=756&forceIcon=1&a`

You're encouraged to play with various sizes, to see the effect it has on the rendering

It's also possible to disable upscaling by adding this to the URL: `&scalingup=0`, but you will need to delete your thumbnails folder in order to be able to get real results, since there is no mechanism in place which detects if a preview has been upscaled or not. We currently only track with or without aspect ratio.
#### Gallery

I also invite you to install [Gallery 8.1](https://apps.owncloud.com/content/show.php?content=169939) to test this since it will generate thumbnails and large previews and would be a good quick test.
Make sure you increase the the max preview dimensions and delete the thumbnails folder prior to testing or your large previews will look tiny.
#### Mobile clients

You should be getting the same previews you were getting before, except maybe a few "fixed" ones.
## Test results
- Pictures should never be bigger than 756x756 (using the config described above)
- Whenever a picture is smaller than what is asked, it will be sent as is if the aspect ratio is required and with a transparent background if no aspect ratio is required.

I've left lots of debugging messages (to be cleanup later), so by looking at the logs, you should be getting an idea of what the class is trying to do with the request it's been given.

### CODE REVIEW

**Comment by oparoz** [comment]
Seems like the CI environment doesn't support movie, office and SVG conversion and my detection routine has failed, so I will update it.

**Comment by oparoz** [comment]
@libasys - Since you've reported the problem, please test this.
Other interested parties: @georgehrke @DeepDiver1975 @MorrisJobke @LukasR... err no ;)

File: tests/lib/preview/provider.php [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+			$this->assertLessThan($this->maxHeight, $preview->height());
+		} elseif ($maxDimRatio > $previewRatio) {
+			$this->assertLessThan($this->maxWidth, $preview->width());
+			$this->assertLessThanOrEqual($this->maxHeight, $preview->height());
+		} else { // Original had to be resized
+			$this->assertLessThanOrEqual($this->maxWidth, $preview->width());
+			$this->assertLessThanOrEqual($this->maxHeight, $preview->height());
+		}
+	}
+}
```
Comment (oparoz) [inline tests/lib/preview/provider.php]: nts: newline

**Comment by oparoz** [comment]
Also, I've introduced some test documents, feel free to give me replacements if you have better ones. I tried to use oC material, but couldn't find a short MP3 and MP4.

**Comment by libasys** [comment]
looks good!

File: lib/private/image.php [hunk 2]
```diff
@@ -973,6 +973,24 @@ public function fitIn($maxWidth, $maxHeight) {
 	}
 
 	/**
+	 * Shrinks the image to fit within a boundary while preserving ratio.
+	 *
+	 * @param integer $maxWidth
+	 * @param integer $maxHeight
+	 * @return bool
+	 */
+	public function fitInIfTooBig($maxWidth, $maxHeight) {
```
Comment (nickvergessen) [inline lib/private/image.php]: missing from public interface?
Also I don't think the name is very suitable.

Comment (oparoz) [inline lib/private/image.php]: Yes, I missed it.

OK about the name.
Let's first validate the concept, then I'll accept any name. I find the original `fitIn` ambiguous because it actually stretches the image to fill the space, so we need a descriptive method name.

Comment (oparoz) [inline lib/private/image.php]: scaleToFit() ?

File: lib/private/image.php [hunk 3]
```diff
@@ -973,6 +973,24 @@ public function fitIn($maxWidth, $maxHeight) {
...
+		$widthOrig = imageSX($this->resource);
+		$heightOrig = imageSY($this->resource);
+
+		if ($widthOrig > $maxWidth || $heightOrig >$maxHeight) {
```
Comment (PVince81) [inline lib/private/image.php]: Should this be `>=` (just asking, I'm not sure myself)

File: lib/private/image.php (hunk 2 above)
Comment (PVince81) [inline lib/private/image.php]: scaleToFit() sounds good. If a small image is stretched to fill the space, this info should be added to the PHPDoc as well.

File: lib/private/image.php (hunk 3 above)
Comment (oparoz) [inline lib/private/image.php]: I don't think so. If it fits in the box already, we don't need to process it further.

**Comment by oparoz** [comment]
I'm going to push updated tests and the fixed preview which goes with it. It fixes a lot of reported issues with previews...

```
PHPUnit 4.5.0 by Sebastian Bergmann and contributors.

................................................................. 65 / 84 ( 77%)
...................

Time: 40.56 seconds, Memory: 34.50Mb

OK (84 tests, 567 assertions)
```

**Comment by oparoz** [comment]
Nice, so if I've done my job right, this shows that the current Preview class fails **51** out of 164 tests.
I'll try to upload my fixed Preview class tomorrow.

**Comment by oparoz** [comment]
OK, so 164 tests were passed and 1127 assertions were verified for the Preview class alone.
I've also performed extensive real world testing and now it's time for you to do the same
@georgehrke @jancborchardt @jospoortvliet @libasys @davivel @setnes @PVince81 @DeepDiver1975 @MorrisJobke @RealRancor @rullzer 

#### Test data
I suggest you upload at least these 4 pictures from the tests/data folder to your test folder:
* testimage.eps | standard eps file, converted by ImageMagick
* testimage.jpg | normal jpg with some test in it
* testimage-wide.png | panoramic picture
* testimageblah.gif | small gif

They were chosen because they trigger different scenarios in the resizing mechanism. You're welcome to test with your libraries as well, just make sure you've deleted the thumbnail folder before proceeding.

Your configuration should have the max preview dimensions set at 756x756 if you want to be able to really test the various restrictions the resizing can encouter.
If you want to test upscaling, then you should have a max_factor of above 1.0 in your configuration

#### Test URLs

**No aspect ratio**
`/index.php/core/preview.png?file=testimage.jpg&x=756&y=756&forceIcon=1`
**With ratio**
`/index.php/core/preview.png?file=testimage.jpg&x=756&y=756&forceIcon=1&a`

You're encouraged to play with various sizes, to see the effect it has on the rendering

It's also possible to disable upscaling by addin this to the URL: `&scalingup=0`, but you will need to delete your thumbnails folder in order to be able to get real results, since there is no mechanism in place which detects if a preview has been upscaled or not. We currently only track with or without aspect ratio.

#### Test results

* Pictures should never be bigger than 756x756 (using the config described above)
* Whenever a picture is smaller than what is asked, it will be sent as is if the aspect ratio is required and with a transparent background if no aspect ratio is required.

Enjoy and hopefully this will be working well for all of you!

I'm also open to discussing testing strategies as this is not easy to do properly and requires large data samples.

File: lib/private/preview.php [hunk 4]
```diff
@@ -95,20 +113,27 @@ public function __construct($user = '', $root = '/', $file = '', $maxX = 1, $max
...
+		$this->configMaxHeight = \OC::$server->getConfig()
+											 ->getSystemValue('preview_max_y', 2048);
+		$this->maxScaleFactor = \OC::$server->getConfig()
+											->getSystemValue('preview_max_scale_factor', 2);
 
 		//save parameters
 		$this->setFile($file);
-		$this->setMaxX($maxX);
-		$this->setMaxY($maxY);
+		$this->setMaxX((int)$maxX);
```
Comment (oparoz) [inline lib/private/preview.php]: The idea is to make sure we get valid dimensions

File: lib/private/preview.php [hunk 5]
```diff
@@ -234,82 +274,91 @@ public function setMimetype($mimeType) {
...
 			throw new \Exception('Cannot set width of 0 or smaller!');
 		}
 		$configMaxX = $this->getConfigMaxX();
-		if (!is_null($configMaxX)) {
-			if ($maxX > $configMaxX) {
-				\OCP\Util::writeLog('core', 'maxX reduced from ' . $maxX . ' to ' . $configMaxX, \OCP\Util::DEBUG);
-				$maxX = $configMaxX;
-			}
-		}
+		$maxX = $this->limitMaxDim($maxX, $configMaxX, 'maxX');
```
Comment (oparoz) [inline lib/private/preview.php]: Extracted into method because of duplication

File: lib/private/preview.php [hunk 6]
```diff
@@ -317,22 +366,28 @@ public function isFileValid() {
...
 	 */
 	public function deletePreview() {
 		$fileInfo = $this->getFileInfo();
-		if($fileInfo !== null && $fileInfo !== false) {
+		if ($fileInfo !== null && $fileInfo !== false) {
 			$fileId = $fileInfo->getId();
 
 			$previewPath = $this->buildCachePath($fileId);
-			return $this->userView->unlink($previewPath);
+			if (!strpos($previewPath, 'max')) {
```
Comment (oparoz) [inline lib/private/preview.php]: Protection to make sure we can never delete the max preview

File: lib/private/preview.php [hunk 7]
```diff
@@ -355,93 +410,243 @@ public function deleteAllPreviews() {
...
 	 */
 	public function isCached($fileId) {
 		if (is_null($fileId)) {
 			return false;
 		}
 
-		// This gives us a calculated path to a preview of asked dimensions
-		// thumbnailFolder/fileId/my_image-<maxX>-<maxY>.png
-		$preview = $this->buildCachePath($fileId);
+		/**
```
Comment (oparoz) [inline lib/private/preview.php]: This whole method has been rewritten so that we know exactly what to look for in the cache. Previous attempt lead to too many misses.

File: lib/private/preview.php [hunk 8]
```diff
@@ -355,93 +410,243 @@ public function deleteAllPreviews() {
...
+		return false;
+	}
+
+	/**
+	 * Returns the dimensions of the max preview
+	 *
+	 * @param FileInfo[] $allThumbnails the list of all our cached thumbnails
+	 *
+	 * @return int[]
+	 */
```
Comment (oparoz) [inline lib/private/preview.php]: Retrieves the dimensions of the max preview by analysing its filename

File: lib/private/preview.php [hunk 9]
```diff
@@ -355,93 +410,243 @@ public function deleteAllPreviews() {
...
 
 	/**
-	 * Checks if a bigger version of a file preview is cached and if not
-	 * return the preview of max allowed dimensions
+	 * Determines the size of the preview we should be looking for in the cache
 	 *
-	 * @param int $fileId fileId of the original image
+	 * @return int[]
+	 */
+	private function simulatePreviewDimensions() {
```
Comment (oparoz) [inline lib/private/preview.php]: New set of methods which determine the final size the preview should have based on the various constraints.

File: lib/private/preview.php [hunk 10]
```diff
@@ -355,93 +410,243 @@ public function deleteAllPreviews() {
...
-
-		$wantedAspectRatio = (float)($this->getMaxX() / $this->getMaxY());
-
 		//array for usable cached thumbnails
 		$possibleThumbnails = array();
-
-		$allThumbnails = $this->userView->getDirectoryContent($previewPath);
 		foreach ($allThumbnails as $thumbnail) {
 			$name = rtrim($thumbnail['name'], '.png');
-			// Always add the max preview to the array
```
Comment (oparoz) [inline lib/private/preview.php]: Removed everything related to the max preview, we only look for larger previews and add the max preview later, if we can't find anything.

File: lib/private/preview.php [hunk 11]
```diff
@@ -563,13 +783,25 @@ private function getCachedPreview($fileId, $cached) {
 
 			$this->preview = $image->valid() ? $image : null;
 
-			$maxX = (int)$this->getMaxX();
-			$maxY = (int)$this->getMaxY();
-			$previewX = (int)$this->preview->width();
-			$previewY = (int)$this->preview->height();
+			if (!is_null($this->preview)) {
+				// Size of the preview we calculated
+				$maxX = $this->previewWidth;
```
Comment (oparoz) [inline lib/private/preview.php]: We compare what we got from the cache with the size of the expected preview, instead of with the requested size.

File: lib/private/preview.php [hunk 12]
```diff
@@ -582,50 +814,135 @@ private function getCachedPreview($fileId, $cached) {
...
+		}
+		$previewWidth = (int)$image->width();
+		$previewHeight = (int)$image->height();
+		$askedWidth = $this->getMaxX();
+		$askedHeight = $this->getMaxY();
+
+		/**
+		 * Phase 1: If required, adjust boundaries to keep aspect ratio
+		 */
+		if ($this->keepAspect) {
```
Comment (oparoz) [inline lib/private/preview.php]: Re-using what we already have in order to avoid duplication

File: lib/private/preview.php [hunk 13]
```diff
@@ -582,50 +814,135 @@ private function getCachedPreview($fileId, $cached) {
...
+			'Your mission is to resize from ' . $previewWidth . 'x' . $previewHeight . ' to '
+			. $askedWidth
+			. 'x' . $askedHeight . ' AR:' . $this->keepAspect, \OCP\Util::DEBUG
+		);
+
+		/**
+		 * Phase 2: Resizes preview to try and match requirements.
+		 * Takes the scaling ratio into consideration
+		 */
+		list($newPreviewWidth, $newPreviewHeight) = $this->scale(
```
Comment (oparoz) [inline lib/private/preview.php]: Everything below this point has been extracted to methods, hoping that one day they'll be migrated to the Image class

File: lib/private/preview.php [hunk 14]
```diff
@@ -639,112 +956,145 @@ private function resizeAndCrop($max = false) {
...
-			$white = imagecolorallocate($backgroundLayer, 255, 255, 255);
-			imagefill($backgroundLayer, 0, 0, $white);
+		if ($previewHeight > $askedHeight) {
+			$cropY = floor(($previewHeight - $askedHeight) * 0.5);
+			$image->crop(0, $cropY, $previewWidth, $askedHeight);
+			$previewHeight = $askedHeight;
+		}
 
-			$image = $image->resource();
+		// Creates a transparent background
```
Comment (oparoz) [inline lib/private/preview.php]: Converted from a white, to a transparent background

File: lib/private/preview.php [hunk 15]
```diff
@@ -639,112 +956,145 @@ private function resizeAndCrop($max = false) {
...
+			$maxX = $this->getMaxX();
+		}
+		if (is_null($maxY)) {
+			$maxY = $this->getMaxY();
+		}
 
 		$previewPath = $this->getPreviewPath($fileId);
 		$previewPath = $previewPath . strval($maxX) . '-' . strval($maxY);
-		if ($this->keepAspect) {
+		$isMaxPreview =
```
Comment (oparoz) [inline lib/private/preview.php]: Some fixes to let us retrieve the max preview when needed

File: lib/private/preview.php [hunk 16]
```diff
@@ -775,49 +1129,51 @@ private function generatePreview($fileId) {
 		$file = $this->getFile();
 		$preview = null;
 
-		$previewProviders = \OC::$server->getPreviewManager()->getProviders();
-			foreach ($previewProviders as $supportedMimeType => $providers) {
-				if (!preg_match($supportedMimeType, $this->mimeType)) {
+		$previewProviders = \OC::$server->getPreviewManager()
```
Comment (oparoz) [inline lib/private/preview.php]: Nothing was touched in this method, it's pure formatting stuff

File: lib/private/preview.php [hunk 17]
```diff
@@ -95,20 +113,27 @@ public function __construct($user = '', $root = '/', $file = '', $maxX = 1, $max
 		$this->userView = new \OC\Files\View('/' . $user);
 
 		//set config
-		$this->configMaxX = \OC::$server->getConfig()->getSystemValue('preview_max_x', 2048);
-		$this->configMaxY = \OC::$server->getConfig()->getSystemValue('preview_max_y', 2048);
-		$this->maxScaleFactor = \OC::$server->getConfig()->getSystemValue('preview_max_scale_factor', 2);
+		$this->configMaxWidth = \OC::$server->getConfig()
```
Comment (nickvergessen) [inline lib/private/preview.php]: `\OC::$server->getConfig()` once and store in a variable?

Comment (oparoz) [inline lib/private/preview.php]: I agree

**Comment by oparoz** [comment]
So, @libasys, did it fix your problems?

@tobiasKaminsky you might be interested in taking a look. It won't fix the thumbnail api issue, but it makes sure that the underlying plumbing is doing its work properly.

**Comment by libasys** [comment]
Should work, i get no errors!

**Comment by oparoz** [comment]
Thanks, now someone familiar with the code should take a look @georgehrke @icewind1991

**Comment by MorrisJobke** [comment]
I had a quick look at the code and it looks good. But I can't judge on all of them.

I tested with several files and aspect ratios and it worked just fine. All files I have here still get nice thumbnails

:+1: from me

**Comment by MorrisJobke** [comment]
@oparoz Can you fix the comment from @nickvergessen ? That would be nice

**Comment by oparoz** [comment]
Thanks for taking the time to test this @MorrisJobke! 
I will use a variable for the $config when I remove all the debug code (unless people want it in...), squash and rebase.

I just need one more :+1: . All the changes have been documented in this convo, so it shouldn't be too difficult to figure out what's going on :)

**Comment by MorrisJobke** [comment]
@DeepDiver1975 Can we get this merged  ... it fixes some issues in the preview handling.

File: lib/private/preview.php [hunk 18]
```diff
@@ -75,18 +87,24 @@ class Preview {
...
 	 * @param bool $scalingUp Disable/Enable upscaling of previews
+	 *
 	 * @throws \Exception
 	 * @return mixed (bool / string)
 	 *                    false if thumbnail does not exist
 	 *                    path to thumbnail if thumbnail exists
 	 */
-	public function __construct($user = '', $root = '/', $file = '', $maxX = 1, $maxY = 1, $scalingUp = true) {
+	public function __construct(
+		$user = '', $root = '/', $file = '', $maxX = 1, $maxY = 1, $scalingUp = true
```
Comment (LukasReschke) [inline lib/private/preview.php]: While squashing please move each to it's own line - THX.

File: lib/private/preview.php (hunk 4 above)
Comment (LukasReschke) [inline lib/private/preview.php]: :rocket:

File: lib/private/preview.php [hunk 19]
```diff
@@ -234,82 +274,91 @@ public function setMimetype($mimeType) {
...
 	/**
-	 * set the the max width of the preview
+	 * Sets the max width of the preview. It's capped by the maximum allowed size set in the
+	 * configuration
+	 *
 	 * @param int $maxX
+	 *
 	 * @throws \Exception
-	 * @return \OC\Preview $this
+	 * @return \OC\Preview
```
Comment (LukasReschke) [inline lib/private/preview.php]: I'd prefer `$this`

File: lib/private/preview.php [hunk 20]
```diff
@@ -234,82 +274,91 @@ public function setMimetype($mimeType) {
...
 	/**
-	 * set the the max height of the preview
+	 * Sets the max height of the preview. It's capped by the maximum allowed size set in the
+	 * configuration
+	 *
 	 * @param int $maxY
+	 *
 	 * @throws \Exception
-	 * @return \OC\Preview $this
+	 * @return \OC\Preview
```
Comment (LukasReschke) [inline lib/private/preview.php]: I'd prefer `$this`

File: lib/private/preview.php [hunk 21]
```diff
@@ -234,82 +274,91 @@ public function setMimetype($mimeType) {
...
 
 	/**
-	 * set whether or not scalingup is enabled
+	 * Sets whether we're allowed to scale up when generating a preview. It's capped by the maximum
+	 * allowed scale factor set in the configuration
+	 *
 	 * @param bool $scalingUp
-	 * @return \OC\Preview $this
+	 *
+	 * @return \OC\Preview
```
Comment (LukasReschke) [inline lib/private/preview.php]: I'd prefer `$this`

File: lib/private/preview.php [hunk 22]
```diff
@@ -234,82 +274,91 @@ public function setMimetype($mimeType) {
...
 		return $this;
 	}
 
 	/**
+	 * Sets whether we need to generate a preview which keeps the aspect ratio of the original file
+	 *
 	 * @param bool $keepAspect
-	 * @return $this
+	 *
+	 * @return \OC\Preview
```
Comment (LukasReschke) [inline lib/private/preview.php]: I'd prefer `$this`

File: lib/private/preview.php [hunk 23]
```diff
@@ -355,93 +410,243 @@ public function deleteAllPreviews() {
...
+			 * Phase 3: We look for a preview of the exact size
+			 */
+			// This gives us a calculated path to a preview of asked dimensions
+			// thumbnailFolder/fileId/<maxX>-<maxY>(-max|-with-aspect).png
+			$preview = $this->buildCachePath($fileId, $previewWidth, $previewHeight);
+
+			// FIXME DEBUG CODE
+			\OCP\Util::writeLog(
+				'core', 'Preview to find in the cache ' . $preview, \OCP\Util::DEBUG
+			);
```
Comment (LukasReschke) [inline lib/private/preview.php]: Just for the record: Will get removed – right?

File: lib/private/preview.php [hunk 24]
```diff
@@ -355,93 +410,243 @@ public function deleteAllPreviews() {
...
+		$maxPreviewX = null;
+		$maxPreviewY = null;
+
+		foreach ($allThumbnails as $thumbnail) {
+			$name = $thumbnail['name'];
+			if (strpos($name, 'max')) {
+				list($maxPreviewX, $maxPreviewY) = $this->getDimensionsFromFilename($name);
+
+				// FIXME DEBUG CODE
+				\OCP\Util::writeLog('core', 'maxPreview ' . $name, \OCP\Util::DEBUG);
```
Comment (LukasReschke) [inline lib/private/preview.php]: As well

File: lib/private/preview.php [hunk 25]
```diff
@@ -355,93 +410,243 @@ public function deleteAllPreviews() {
...
+	}
+
+	/**
+	 * Returns the dimensions of the max preview
+	 *
+	 * @param FileInfo[] $allThumbnails the list of all our cached thumbnails
+	 *
+	 * @return int[]
+	 */
+	private function getMaxPreviewSize($allThumbnails) {
```
Comment (LukasReschke) [inline lib/private/preview.php]: Can we type hint `array` here?

File: lib/private/preview.php [hunk 26]
```diff
@@ -355,93 +410,243 @@ public function deleteAllPreviews() {
...
+		if ($askedWidth / $originalRatio < $askedHeight) {
+			// width restricted
+			$askedHeight = round($askedWidth / $originalRatio);
+		} else {
+			$askedWidth = round($askedHeight * $originalRatio);
+		}
 
-		if (is_null($fileId)) {
-			return false;
+		return [(int)$askedWidth, (int)$askedHeight];
```
Comment (LukasReschke) [inline lib/private/preview.php]: I really like how you cast stuff. Kudos for that :rocket: :beers

File: lib/private/preview.php [hunk 27]
```diff
@@ -355,93 +410,243 @@ public function deleteAllPreviews() {
...
+	}
+
+	private function fixSize($askedWidth, $askedHeight) {
```
Comment (LukasReschke) [inline lib/private/preview.php]: Can we haz :cookie: and PHPDoc?

File: lib/private/preview.php [hunk 28]
```diff
@@ -355,93 +410,243 @@ public function deleteAllPreviews() {
...
+	/**
+	 * Checks if a bigger version of a file preview is cached and if not
+	 * return the preview of max allowed dimensions
+	 *
+	 * @param int $fileId fileId of the original image
+	 * @param FileInfo[] $allThumbnails the list of all our cached thumbnails
+	 *
+	 * @return string path to bigger thumbnail
+	 */
+	private function isCachedBigger($fileId, $allThumbnails) {
```
Comment (LukasReschke) [inline lib/private/preview.php]: can we typehint `array` here on `$allThumbnails`?

File: lib/private/preview.php [hunk 29]
```diff
@@ -355,93 +410,243 @@ public function deleteAllPreviews() {
...
-		// At this stage, we didn't find a preview, so if the folder is not empty,
-		// we return the max preview we generated on the first run
-		if ($possibleThumbnails) {
-			return $possibleThumbnails['max'];
-		}
+		// FIXME DEBUG CODE
+		\OCP\Util::writeLog(
+			'core', "Bigger preview not found, so returning max preview",
+			\OCP\Util::DEBUG
+		);
```
Comment (LukasReschke) [inline lib/private/preview.php]: As well

File: lib/private/preview.php [hunk 30]
```diff
@@ -355,93 +410,243 @@ public function deleteAllPreviews() {
...
-
-		if (is_null($fileId)) {
-			return array();
+	 * Get possible bigger thumbnails of the given image with the proper aspect ratio
+	 *
+	 * @param FileInfo[] $allThumbnails the list of all our cached thumbnails
+	 *
+	 * @return string[] an array of paths to bigger thumbnails
+	 */
+	private function getPossibleThumbnails($allThumbnails) {
```
Comment (LukasReschke) [inline lib/private/preview.php]: can we typehint `array`?

File: lib/private/preview.php [hunk 31]
```diff
@@ -556,6 +770,12 @@ public function showPreview($mimeType = null) {
 	 */
 	private function getCachedPreview($fileId, $cached) {
 		$stream = $this->userView->fopen($cached, 'r');
+
+		// FIXME DEBUG CODE
+		\OCP\Util::writeLog(
+			'core', 'We are looking for this thumbnail: ' . $cached, \OCP\Util::DEBUG
+		);
```
Comment (LukasReschke) [inline lib/private/preview.php]: As well

File: lib/private/preview.php [hunk 32]
```diff
@@ -563,13 +783,25 @@ private function getCachedPreview($fileId, $cached) {
...
+
+				// We don't have an exact match
+				if ($previewX !== $maxX || $previewY !== $maxY) {
+					$this->resizeAndStore($fileId);
+				} else {
+
+					// FIXME DEBUG CODE
+					\OCP\Util::writeLog(
+						'core', 'We have found the correct thumbnail', \OCP\Util::DEBUG
+					);
```
Comment (LukasReschke) [inline lib/private/preview.php]: As well

File: lib/private/preview.php [hunk 33]
```diff
@@ -582,50 +814,135 @@ private function getCachedPreview($fileId, $cached) {
...
-	/**
-	 * resize, crop and fix orientation
-	 *
-	 * @param bool $max
-	 */
-	private function resizeAndCrop($max = false) {
 		$image = $this->preview;
+		if (!($image instanceof \OCP\IImage)) {
+			\OCP\Util::writeLog(
+				'core', '$this->preview is not an instance of OC_Image', \OCP\Util::DEBUG
```
Comment (LukasReschke) [inline lib/private/preview.php]: `OC_Image` or `\OCP\IImage`?

File: lib/private/preview.php (hunk 18 above)
Comment (oparoz) [inline lib/private/preview.php]: ✓

File: lib/private/preview.php (hunk 19 above)
Comment (oparoz) [inline lib/private/preview.php]: Any particular reason? $this is not a datatype.

File: lib/private/preview.php (hunk 23 above)
Comment (oparoz) [inline lib/private/preview.php]: Yes, I'll keep a copy in my records, in case this needs to be revisited in the future.

File: lib/private/preview.php (hunk 25 above)
Comment (oparoz) [inline lib/private/preview.php]: It's already type hinted, unless you mean to declare the type of everything inside a FileInfo object? I don't think that's ever been documented.

File: lib/private/preview.php (hunk 27 above)
Comment (oparoz) [inline lib/private/preview.php]: ✓

File: lib/private/preview.php (hunk 33 above)
Comment (oparoz) [inline lib/private/preview.php]: ✓

File: lib/private/preview.php (hunk 17 above)
Comment (oparoz) [inline lib/private/preview.php]: ✓

**Comment by oparoz** [comment]
Thanks for the review @LukasReschke :beers: 
I've squashed + rebased and will add another commit once I understand your preference for $this in PHPDocs and what you need in terms of type hinting for arrays which are already typed.

**Comment by MorrisJobke** [comment]
@LukasReschke Is this now okay from your side?

File: lib/private/image.php [hunk 34]
```diff
@@ -973,6 +975,24 @@ public function fitIn($maxWidth, $maxHeight) {
...
 
 	/**
+	 * Shrinks larger images to fit within specified boundaries while preserving ratio.
+	 *
+	 * @param integer $maxWidth
+	 * @param integer $maxHeight
+	 * @return bool
+	 */
+	public function scaleDownToFit($maxWidth, $maxHeight) {
+		$widthOrig = imageSX($this->resource);
```
Comment (DeepDiver1975) [inline lib/private/image.php]: test if $this->resource is valid is missing here - use 

```php
		if (!$this->valid()) {
			$this->logger->error(__METHOD__ . '(): No image loaded', array('app' => 'core'));
			return false;
		}

Comment (oparoz) [inline lib/private/image.php]: I removed this on purpose since we either return false or call `fitIn` which will make sure there is an image to work on before trying to process it, but I can see now why it's still a good idea to have it since the method is first determining the size of the resource in order to make a decision.

Comment (oparoz) [inline lib/private/image.php]: Or should I modify the original method in this PR to avoid duplication?
`public function fitIn($maxWidth, $maxHeight, $scalingUp = false)`

Comment (DeepDiver1975) [inline lib/private/image.php]: > I think in a future version of oC, we could even merge the two methods by adding an extra argument. I didn't want to touch the original this late in the game.

I'd walk this path for now - THX

**Comment by DeepDiver1975** [comment]
:+1: from my side - THX a lot!

**Comment by jospoortvliet** [comment]
Yes, super awesome @oparoz thanks a bunch!
